File: app/game_engine/boards/hexagonal_board.rb
class HexagonalBoard

  attr_reader :size

  def initialize(size)
    @size = size
  end

  def coordinate_valid?(coordinate)
    sum = coordinate['x'] + coordinate['y']

    coordinate['x'].between?(0, 2 * size) &&
    coordinate['y'].between?(0, 2 * size) &&
    sum.between?(size, size + 2 * size)
  end

  def center_coordinate
    { 'x' => size, 'y' => size }
  end

  def territory(coordinate)
    if coordinate['y'] == size
      'neutral'
    elsif coordinate['y'] < size
      'alabaster'
    else
      'onyx'
    end
  end

  def directional_functions(type)
    if type == 'orthogonal'
      [
        lambda{ |coordinate| coordinate['x'] += 1},
        lambda{ |coordinate| coordinate['x'] -= 1},
        lambda{ |coordinate| coordinate['y'] += 1},
        lambda{ |coordinate| coordinate['y'] -= 1},
        lambda{ |coordinate| coordinate['x'] += 1; coordinate['y'] -= 1},
        lambda{ |coordinate| coordinate['x'] -= 1; coordinate['y'] += 1}
      ]
    elsif type == 'diagonal'
      [
        lambda{ |coordinate| coordinate['x'] += 1; coordinate['y'] += 1},
        lambda{ |coordinate| coordinate['x'] -= 1; coordinate['y'] -= 1},
        lambda{ |coordinate| coordinate['x'] += 1; coordinate['y'] -= 2},
        lambda{ |coordinate| coordinate['x'] -= 1; coordinate['y'] += 2},
        lambda{ |coordinate| coordinate['x'] += 2; coordinate['y'] -= 1},
        lambda{ |coordinate| coordinate['x'] -= 2; coordinate['y'] += 1}
      ]
    else
      raise "#{self.class}#directional_functions does not support type: #{type}"
    end
  end

  def distance(coordinate1, coordinate2)
    x_diff = coordinate1['x'] - coordinate2['x']
    y_diff = coordinate1['y'] - coordinate2['y']
    d_diff = x_diff + y_diff

    [x_diff, y_diff, d_diff].map(&:abs).max
  end
end

File: api/src/services/game/board/hexagonal_board.ts
import { ICoordinate } from "../../../shared/dtos/game";
import {
  ICoordinateUpdater,
  BoardDirection,
  SetupTerritoryOwner,
  IBoard,
} from ".";

interface IRange {
  min: number;
  max: number;
}

export class HexagonalBoard implements IBoard {
  private readonly size: number;

  constructor(size: number) {
    this.size = size;
  }

  getAllCoordinates(): ICoordinate[] {
    const result: ICoordinate[] = [];
    const range = this.getCoordinateRange();
    for (let x = range.min; x <= range.max; x++) {
      for (let y = range.min; y <= range.max; y++) {
        const coordinate = { x, y };
        if (this.isCoordinateValid(coordinate)) {
          result.push(coordinate);
        }
      }
    }
    return result;
  }

  getCenter(): ICoordinate {
    return { x: 0, y: 0 };
  }

  getCoordinateDistance(
    coordinate1: ICoordinate,
    coordinate2: ICoordinate
  ): number {
    const xDiff = coordinate1.x - coordinate2.x;
    const yDiff = coordinate1.y - coordinate2.y;
    const diagonalDiff = xDiff + yDiff;
    return Math.max(...[xDiff, yDiff, diagonalDiff].map((x) => Math.abs(x)));
  }

  getDirectionalFunctions(type: BoardDirection): ICoordinateUpdater[] {
    if (type === BoardDirection.orthogonal) {
      return [
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y }),
        (coordinate) => ({ x: coordinate.x, y: coordinate.y + 1 }),
        (coordinate) => ({ x: coordinate.x, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y + 1 }),
      ];
    }
    if (type === BoardDirection.diagonal) {
      return [
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y + 1 }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x + 1, y: coordinate.y - 2 }),
        (coordinate) => ({ x: coordinate.x - 1, y: coordinate.y + 2 }),
        (coordinate) => ({ x: coordinate.x + 2, y: coordinate.y - 1 }),
        (coordinate) => ({ x: coordinate.x - 2, y: coordinate.y + 1 }),
      ];
    }
    throw Error("Unsupported board direction");
  }

  getSetupTerritoryOwner(coordinate: ICoordinate): SetupTerritoryOwner {
    if (coordinate.y === 0) {
      return SetupTerritoryOwner.neutral;
    }
    if (coordinate.y < 0) {
      return SetupTerritoryOwner.alabaster;
    }
    return SetupTerritoryOwner.onyx;
  }

  isCoordinateValid(coordinate: ICoordinate): boolean {
    const range = this.getCoordinateRange();
    const sum = coordinate.x + coordinate.y;
    return (
      this.isBetween(coordinate.x, range.min, range.max) &&
      this.isBetween(coordinate.y, range.min, range.max) &&
      this.isBetween(sum, range.min, range.max)
    );
  }

  private isBetween(value: number, min: number, max: number): boolean {
    return value >= min && value <= max;
  }

  private getCoordinateRange(): IRange {
    return { min: -1 * this.size, max: this.size };
  }
}
